### PR TITLE
[FLINK-30268] HA metadata and other cluster submission related errors should not throw DeploymentFailedException

### PR DESCRIPTION
**Pull request by pvary**
## What is the purpose of the change

Make it possible to recover from the missing HA errors

## Brief change log

- Created a new Exception signalling recoverable errors
- Made sure that the state of the cluster/job is not changed when a recoverable error happens
- Added unit test

## Verifying this change
This change added tests and can be verified as follows:
- Added a unit test recovering from missing HA

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by mbalassi** [comment]
Thanks @pvary. `DeploymentRecoveryTest#verifyApplicationJmRecovery` is failing, PTAL on Monday.

**Comment by gyfora** [comment]
I would not necessarily call missing HA metadata exceptions recoverable. In most cases they are actually fatal and as many error msgs suggest, they require deletion and recreation of the CR.

**Comment by morhidi** [comment]
@pvary please also rebase it to the latest containing @gaborgsomogyi's e2e changes

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java [hunk 1]
```diff
@@ -139,10 +142,13 @@ public UpdateControl<FlinkDeployment> reconcile(FlinkDeployment flinkApp, Contex
                 configManager.getOperatorConfiguration(), flinkApp, previousDeployment, true);
     }
 
-    private void handleDeploymentFailed(FlinkDeployment flinkApp, DeploymentFailedException dfe) {
+    private void handleDeploymentFailed(
+            FlinkDeployment flinkApp, DeploymentFailedException dfe, boolean updateJobStatus) {
         LOG.error("Flink Deployment failed", dfe);
-        flinkApp.getStatus().setJobManagerDeploymentStatus(JobManagerDeploymentStatus.ERROR);
-        flinkApp.getStatus().getJobStatus().setState(JobStatus.RECONCILING.name());
+        if (updateJobStatus) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:148]: These should be two separate methods, the update status flag does not have any well defined meaning I feel

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/exception/RecoverableDeploymentFailureException.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.operator.exception;
+
+/** Exception to signal non-terminal deployment failure. */
+public class RecoverableDeploymentFailureException extends DeploymentFailedException {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/exception/RecoverableDeploymentFailureException.java:21]: Let’s not extend DeploymentFailedException , the deployment does not fail in these cases (the jobmanager deployment I mean)

we could simply call FlinkResourceException or something more generic

**Review by gyfora — CHANGES_REQUESTED** [review]

**Comment by pvary** [comment]
> @pvary please also rebase it to the latest containing @gaborgsomogyi's e2e changes

@morhidi: I am not sure I understand this comment. You might refer to my refactoring of the e2e tests which groups them to 8 executor. If this is the case, then on Friday we had to revert it, because it caused continuous failures. The decision was to revert it for the release and find fix it when we have more time.

**Comment by morhidi** [comment]
> > @pvary please also rebase it to the latest containing @gaborgsomogyi's e2e changes
> 
> @morhidi: I am not sure I understand this comment. You might refer to my refactoring of the e2e tests which groups them to 8 executor. If this is the case, then on Friday we had to revert it, because it caused continuous failures. The decision was to revert it for the release and find fix it when we have more time.

@gaborgsomogyi has added another e2e related fix since then.

**Review by gyfora — APPROVED** [review]

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java [hunk 3]
```diff
@@ -125,7 +125,7 @@ protected Optional<UpgradeMode> getAvailableUpgradeMode(
 
         if (status.getJobManagerDeploymentStatus() == JobManagerDeploymentStatus.MISSING
                 || status.getJobManagerDeploymentStatus() == JobManagerDeploymentStatus.ERROR) {
-            throw new DeploymentFailedException(
+            throw new RecoveryFailureException(
                     "JobManager deployment is missing and HA data is not available to make stateful upgrades. "
                             + "It is possible that the job has finished or terminally failed, or the configmaps have been deleted. "
                             + "Manual restore required.",
```
Comment (gaborgsomogyi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java:131]: How true is this? `Manual restore required.`

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java (hunk 3 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java:131]: This almost certainly means that the operator cannot restore this anymore. The user needs to delete the CR and recover using initialSavepoontPath at this stage

**Comment by gaborgsomogyi** [comment]
Still there are errors:
```
Error:  Errors: 
Error:    ApplicationReconcilerTest.testUpgrade:211 » RecoveryFailure JobManager deploym...
Error:    ApplicationReconcilerTest.testUpgrade:211 » RecoveryFailure JobManager deploym...
Error:    ApplicationReconcilerTest.testUpgrade:211 » RecoveryFailure JobManager deploym...
Error:    ApplicationReconcilerTest.testUpgrade:211 » RecoveryFailure JobManager deploym...
Error:    NativeFlinkServiceTest.testGetLastCheckpoint:287 » RecoveryFailure Latest chec...
[INFO] 
Error:  Tests run: 326, Failures: 0, Errors: 5, Skipped: 0
```

**Comment by mbalassi** [comment]
I think we have reached consensus on this for the 1.3 release, so I a merging.

**Pull request merged**